Silent MSBuild task crash

```
func read_coords(): Tuple<int32, int32> {
    val line = ReadLine();
    return Tuple(int32.Parse("\{line[0]}"), int32.Parse("\{line[2]}"));
}
```

error:
```
C:\Development\DracoTest\Toolchain\GlobalPackages\draco.compiler.toolset\1.0.0\build\Draco.Compiler.Toolset.targets(11,
5): error MSB4181: The "DracoCompiler" task returned false but did not log an error. [C:\Development\DracoTest\DracoTes
t.dracoproj]

The build failed. Fix the build errors and run again.
```

--- a/src/Draco.Compiler.Tests/EndToEnd/BclUsageTests.cs
+++ b/src/Draco.Compiler.Tests/EndToEnd/BclUsageTests.cs
@@ -324,4 +324,22 @@ public sealed class BclUsageTests : EndToEndTestsBase
 
         Assert.Equal(123, result);
     }
+
+    [Fact]
+    public void StringSplitting()
+    {
+        var assembly = Compile("""
+            import System;
+
+            public func foo(): string {
+                val parts = "1, 2, 3".Split(",", StringSplitOptions.TrimEntries);
+                return string.Join(";", parts);
+            }
+            """);
+        var result = Invoke<string>(
+            assembly: assembly,
+            methodName: "foo");
+
+        Assert.Equal("1;2;3", result);
+    }
 }

--- a/src/Draco.Compiler.Tests/EndToEnd/CompilingCodeTests.cs
+++ b/src/Draco.Compiler.Tests/EndToEnd/CompilingCodeTests.cs
@@ -672,4 +672,15 @@ public sealed class CompilingCodeTests : EndToEndTestsBase
         var x = Invoke<int>(assembly, "get_result");
         Assert.Equal(24, x);
     }
+
+    [Fact]
+    public void SingleInterpolatedElement()
+    {
+        var assembly = Compile(""""
+            public func stringify(a: int32): string = "\{a}";
+            """");
+
+        var x = Invoke<string>(assembly, "stringify", 123);
+        Assert.Equal("123", x);
+    }
 }

--- a/src/Draco.Compiler.Tests/EndToEnd/Issue139Tests.cs
+++ b/src/Draco.Compiler.Tests/EndToEnd/Issue139Tests.cs
@@ -194,6 +194,9 @@ public sealed class Issue139Tests
         """")]
     [InlineData("public import Foo")]
     [InlineData("label:")]
+    [InlineData("""
+        func foo(): string = "\{0}";
+        """)]
     [Theory]
     public void DoesNotCrash(string source)
     {

--- a/src/Draco.Compiler.Tests/Semantics/TypeCheckingTests.cs
+++ b/src/Draco.Compiler.Tests/Semantics/TypeCheckingTests.cs
@@ -1940,7 +1940,7 @@ public sealed class TypeCheckingTests : SemanticTestsBase
 
         // Assert
         Assert.Single(diags);
-        AssertDiagnostic(diags, TypeCheckingErrors.NoMatchingOverload);
+        AssertDiagnostic(diags, TypeCheckingErrors.TypeMismatch);
     }
 
     [Fact]

--- a/src/Draco.Compiler.Tests/Semantics/TypeCheckingTests.cs
+++ b/src/Draco.Compiler.Tests/Semantics/TypeCheckingTests.cs
@@ -1,4 +1,3 @@
-using System;
 using System.Collections.Immutable;
 using Draco.Compiler.Api;
 using Draco.Compiler.Api.Syntax;

--- a/src/Draco.Compiler.Tests/Semantics/TypeCheckingTests.cs
+++ b/src/Draco.Compiler.Tests/Semantics/TypeCheckingTests.cs
@@ -1,3 +1,4 @@
+using System;
 using System.Collections.Immutable;
 using Draco.Compiler.Api;
 using Draco.Compiler.Api.Syntax;
@@ -2141,5 +2142,42 @@ public sealed class TypeCheckingTests : SemanticTestsBase
         Assert.True(SymbolEqualityComparer.Default.Equals(intArrayType, aSym.Type));
         Assert.True(SymbolEqualityComparer.Default.Equals(intArrayType, bSym.Type));
         Assert.True(SymbolEqualityComparer.Default.Equals(intArrayType, tmpSym.Type));
+    }
+
+    [Fact]
+    public void UndexerWitingOverloadedCall()
+    {
+        // import System;
+        // func read_coords(): int32 {
+        //     val line = Console.ReadLine();
+        //     return int32.Parse(line[0]);
+        // }
+
+        var main = SyntaxTree.Create(CompilationUnit(
+            ImportDeclaration("System"),
+            FunctionDeclaration(
+                "read_coords",
+                ParameterList(),
+                NameType("int32"),
+                BlockFunctionBody(
+                    DeclarationStatement(ImmutableVariableDeclaration(
+                        "line",
+                        null,
+                        CallExpression(MemberExpression(NameExpression("Console"), "ReadLine")))),
+                    ExpressionStatement(ReturnExpression(
+                        CallExpression(
+                            MemberExpression(NameExpression("int32"), "Parse"),
+                            IndexExpression(NameExpression("line"), LiteralExpression(0)))))))));
+
+        // Act
+        var compilation = CreateCompilation(main);
+
+        var semanticModel = compilation.GetSemanticModel(main);
+
+        var diags = semanticModel.Diagnostics;
+
+        // Assert
+        Assert.Single(diags);
+        AssertDiagnostic(diags, TypeCheckingErrors.TypeMismatch);
     }
 }

--- a/src/Draco.Compiler.Tests/Semantics/TypeCheckingTests.cs
+++ b/src/Draco.Compiler.Tests/Semantics/TypeCheckingTests.cs
@@ -2145,7 +2145,7 @@ public sealed class TypeCheckingTests : SemanticTestsBase
     }
 
     [Fact]
-    public void UndexerWitingOverloadedCall()
+    public void IndexerWitingOverloadedCall()
     {
         // import System;
         // func read_coords(): int32 {

--- a/src/Draco.Compiler/Internal/Binding/BinderFacts.cs
+++ b/src/Draco.Compiler/Internal/Binding/BinderFacts.cs
@@ -170,8 +170,32 @@ internal static class BinderFacts
         case bool:
             type = intrinsics.Bool;
             return true;
+        case byte:
+            type = intrinsics.Uint8;
+            return true;
+        case ushort:
+            type = intrinsics.Uint16;
+            return true;
+        case uint:
+            type = intrinsics.Uint32;
+            return true;
+        case ulong:
+            type = intrinsics.Uint64;
+            return true;
+        case sbyte:
+            type = intrinsics.Int8;
+            return true;
+        case short:
+            type = intrinsics.Int16;
+            return true;
         case int:
             type = intrinsics.Int32;
+            return true;
+        case long:
+            type = intrinsics.Int64;
+            return true;
+        case float:
+            type = intrinsics.Float32;
             return true;
         case double:
             type = intrinsics.Float64;

--- a/src/Draco.Compiler/Internal/Binding/Binder_UntypedExpression.cs
+++ b/src/Draco.Compiler/Internal/Binding/Binder_UntypedExpression.cs
@@ -251,7 +251,7 @@ internal partial class Binder
             {
                 // Retry binding with the resolved variant
                 var call = this.BindCallExpression(syntax, delayed.Promise.Result, args, constraints, diagnostics);
-                constraints.Unify(promisedType, call.TypeRequired);
+                constraints.UnifyAsserted(promisedType, call.TypeRequired);
                 return call;
             });
             return new UntypedDelayedExpression(syntax, promise, promisedType);
@@ -296,7 +296,7 @@ internal partial class Binder
                     symbolPromise.ConfigureDiagnostic(diag => diag
                         .WithLocation(syntax.Function.Location));
 
-                    constraints.Unify(resultType, promisedType);
+                    constraints.UnifyAsserted(resultType, promisedType);
                     return new UntypedCallExpression(syntax, mem.Accessed, symbolPromise, args, resultType);
                 }
                 else
@@ -308,7 +308,7 @@ internal partial class Binder
                     callPromise.ConfigureDiagnostic(diag => diag
                         .WithLocation(syntax.Location));
 
-                    constraints.Unify(resultType, promisedType);
+                    constraints.UnifyAsserted(resultType, promisedType);
                     return new UntypedIndirectCallExpression(syntax, mem, args, resultType);
                 }
             });
@@ -531,14 +531,14 @@ internal partial class Binder
                     template: SymbolResolutionErrors.NoGettableIndexerInType,
                     location: index.Location,
                     formatArgs: receiver.Type));
-                constraints.Unify(returnType, IntrinsicSymbols.ErrorType);
+                constraints.UnifyAsserted(returnType, IntrinsicSymbols.ErrorType);
                 return ConstraintPromise.FromResult<FunctionSymbol>(new NoOverloadFunctionSymbol(args.Length));
             }
             var overloaded = constraints.Overload(
                 indexers,
                 args.Cast<object>().ToImmutableArray(),
                 out var gotReturnType);
-            constraints.Unify(returnType, gotReturnType);
+            constraints.UnifyAsserted(returnType, gotReturnType);
             overloaded.ConfigureDiagnostic(diag => diag
                 .WithLocation(index.Location));
             return overloaded;

--- a/src/Draco.Compiler/Internal/Binding/Binder_UntypedLvalue.cs
+++ b/src/Draco.Compiler/Internal/Binding/Binder_UntypedLvalue.cs
@@ -129,7 +129,7 @@ internal partial class Binder
                     template: SymbolResolutionErrors.NoSettableIndexerInType,
                     location: index.Location,
                     formatArgs: receiverType));
-                constraints.Unify(returnType, IntrinsicSymbols.ErrorType);
+                constraints.UnifyAsserted(returnType, IntrinsicSymbols.ErrorType);
                 return ConstraintPromise.FromResult<FunctionSymbol>(new NoOverloadFunctionSymbol(args.Length + 1));
             }
             var overloaded = constraints.Overload(

--- a/src/Draco.Compiler/Internal/Codegen/CilCodegen.cs
+++ b/src/Draco.Compiler/Internal/Codegen/CilCodegen.cs
@@ -248,17 +248,8 @@ internal sealed class CilCodegen
             foreach (var index in storeElement.Indices) this.EncodePush(index);
             this.EncodePush(storeElement.Source);
 
-            var storedValueType = storeElement.Source.Type!.Substitution;
             // TODO: Not the prettiest...
             var targetStorageType = storeElement.TargetArray.Type!.Substitution.GenericArguments[0].Substitution;
-            var needsToBox = storedValueType.IsValueType && !targetStorageType.IsValueType;
-
-            if (needsToBox)
-            {
-                // We need to box it
-                this.InstructionEncoder.OpCode(ILOpCode.Box);
-                this.EncodeToken(storeElement.Source.Type!);
-            }
 
             if (storeElement.Indices.Count == 1)
             {

--- a/src/Draco.Compiler/Internal/Codegen/CilCodegen.cs
+++ b/src/Draco.Compiler/Internal/Codegen/CilCodegen.cs
@@ -344,6 +344,17 @@ internal sealed class CilCodegen
             this.StoreLocal(arrLen.Target);
             break;
         }
+        case BoxInstruction box:
+        {
+            // Value to be boxed
+            this.EncodePush(box.Value);
+            // Box it
+            this.InstructionEncoder.OpCode(ILOpCode.Box);
+            this.EncodeToken(box.Value.Type!);
+            // Sore result
+            this.StoreLocal(box.Target);
+            break;
+        }
         default:
             throw new ArgumentOutOfRangeException(nameof(instruction));
         }

--- a/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
+++ b/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
@@ -353,8 +353,10 @@ internal partial class LocalRewriter : BoundTreeRewriter
         {
             // Lower the expression
             var arg = (BoundExpression)singleInterpolation.Value.Accept(this);
-            // TODO: Just call ToString on it
-            throw new System.NotImplementedException();
+            return CallExpression(
+                receiver: arg,
+                method: this.WellKnownTypes.SystemObject_ToString,
+                arguments: ImmutableArray<BoundExpression>.Empty);
         }
         // We need to desugar into string.Format("format string", array of args)
         // Build up interpolation string and lower interpolated expressions

--- a/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
+++ b/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
@@ -426,20 +426,35 @@ internal partial class LocalRewriter : BoundTreeRewriter
 
     public override BoundNode VisitPropertySetExpression(BoundPropertySetExpression node)
     {
-        // property = x
+        // property = expr
         //
         // =>
         //
-        // property_set(x)
+        // {
+        //     var tmp = expr;
+        //     property_set(tmp)
+        //     expr
+        // }
 
         var receiver = node.Receiver is null ? null : (BoundExpression)node.Receiver.Accept(this);
         var setter = node.Setter;
         var value = (BoundExpression)node.Value.Accept(this);
 
-        return CallExpression(
-            receiver: receiver,
-            method: setter,
-            arguments: ImmutableArray.Create(value));
+        var tmp = this.StoreTemporary(value);
+
+        var result = BlockExpression(
+            locals: tmp.Symbol is null
+                ? ImmutableArray<LocalSymbol>.Empty
+                : ImmutableArray.Create(tmp.Symbol),
+            statements: ImmutableArray.Create(
+                tmp.Assignment,
+                ExpressionStatement(CallExpression(
+                    receiver: receiver,
+                    method: setter,
+                    arguments: ImmutableArray.Create(value)))),
+            value: tmp.Reference);
+
+        return result.Accept(this);
     }
 
     public override BoundNode VisitPropertyGetExpression(BoundPropertyGetExpression node)

--- a/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
+++ b/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
@@ -433,7 +433,7 @@ internal partial class LocalRewriter : BoundTreeRewriter
         // {
         //     var tmp = expr;
         //     property_set(tmp)
-        //     expr
+        //     tmp
         // }
 
         var receiver = node.Receiver is null ? null : (BoundExpression)node.Receiver.Accept(this);
@@ -451,7 +451,7 @@ internal partial class LocalRewriter : BoundTreeRewriter
                 ExpressionStatement(CallExpression(
                     receiver: receiver,
                     method: setter,
-                    arguments: ImmutableArray.Create(value)))),
+                    arguments: ImmutableArray.Create(tmp.Reference)))),
             value: tmp.Reference);
 
         return result.Accept(this);

--- a/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
+++ b/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
@@ -432,7 +432,7 @@ internal partial class LocalRewriter : BoundTreeRewriter
         //
         // {
         //     var tmp = expr;
-        //     property_set(tmp)
+        //     property_set(tmp);
         //     tmp
         // }
 

--- a/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
+++ b/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
@@ -353,6 +353,7 @@ internal partial class LocalRewriter : BoundTreeRewriter
         {
             // Lower the expression
             var arg = (BoundExpression)singleInterpolation.Value.Accept(this);
+            // TODO: We need to BOX here if it's a value-type
             return CallExpression(
                 receiver: arg,
                 method: this.WellKnownTypes.SystemObject_ToString,

--- a/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
+++ b/src/Draco.Compiler/Internal/Lowering/LocalRewriter.cs
@@ -353,7 +353,6 @@ internal partial class LocalRewriter : BoundTreeRewriter
         {
             // Lower the expression
             var arg = (BoundExpression)singleInterpolation.Value.Accept(this);
-            // TODO: We need to BOX here if it's a value-type
             return CallExpression(
                 receiver: arg,
                 method: this.WellKnownTypes.SystemObject_ToString,

--- a/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
@@ -82,14 +82,26 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
         return proc;
     }
 
+    private static bool NeedsBoxing(TypeSymbol targetType, TypeSymbol sourceType)
+    {
+        var targetIsValueType = targetType.Substitution.IsValueType;
+        var sourceIsValueType = sourceType.Substitution.IsValueType;
+        return !targetIsValueType && sourceIsValueType;
+    }
+
+    private static bool NeedsUnboxing(TypeSymbol targetType, TypeSymbol sourceType)
+    {
+        var targetIsValueType = targetType.Substitution.IsValueType;
+        var sourceIsValueType = sourceType.Substitution.IsValueType;
+        return targetIsValueType && !sourceIsValueType;
+    }
+
     public IOperand BoxIfNeeded(TypeSymbol targetType, IOperand source)
     {
         if (source.Type is null) throw new System.ArgumentException("source must be a typed operand", nameof(source));
 
-        var targetIsValueType = targetType.Substitution.IsValueType;
-        var sourceIsValueType = source.Type.Substitution.IsValueType;
-        var needsBox = !targetIsValueType && sourceIsValueType;
-        var needsUnbox = targetIsValueType && !sourceIsValueType;
+        var needsBox = NeedsBoxing(targetType, source.Type);
+        var needsUnbox = NeedsUnboxing(targetType, source.Type);
 
         if (needsBox)
         {
@@ -270,12 +282,21 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
         }
         else
         {
-            var receiver = node.Receiver.TypeRequired.IsValueType
-                ? this.CompileToAddress(node.Receiver)
-                : this.Compile(node.Receiver);
-            if (node.Method.ContainingSymbol is TypeSymbol methodContainer)
+            IOperand receiver;
+            // Box receiver, if needed
+            if (node.Method.ContainingSymbol is TypeSymbol methodContainer
+             && NeedsBoxing(methodContainer, node.Receiver.TypeRequired))
             {
-                receiver = this.BoxIfNeeded(methodContainer, receiver);
+                receiver = this.Compile(node.Receiver);
+                var newReceiver = this.DefineRegister(methodContainer);
+                this.Write(Box(newReceiver, methodContainer, receiver));
+                receiver = newReceiver;
+            }
+            else
+            {
+                receiver = node.Receiver.TypeRequired.IsValueType
+                    ? this.CompileToAddress(node.Receiver)
+                    : this.Compile(node.Receiver);
             }
             var args = node.Arguments
                 .Zip(node.Method.Parameters)

--- a/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
@@ -267,7 +267,6 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
 
     public override IOperand VisitCallExpression(BoundCallExpression node)
     {
-        var isSetter = node.Method is IPropertyAccessorSymbol p && p.Property.Setter == node.Method;
         // TODO: Lots of duplication, we should merge these instructions...
         var receiver = this.CompileReceiver(node);
         var args = node.Arguments
@@ -284,7 +283,7 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
         {
             this.Write(MemberCall(callResult, proc, receiver, args));
         }
-        return isSetter ? args[^1] : callResult;
+        return callResult;
     }
 
     private IOperand? CompileReceiver(BoundCallExpression call)

--- a/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
@@ -269,43 +269,41 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
     {
         var isSetter = node.Method is IPropertyAccessorSymbol p && p.Property.Setter == node.Method;
         // TODO: Lots of duplication, we should merge these instructions...
-        if (node.Receiver is null)
+        var receiver = this.CompileReceiver(node);
+        var args = node.Arguments
+            .Zip(node.Method.Parameters)
+            .Select(pair => this.BoxIfNeeded(pair.Second.Type, this.Compile(pair.First)))
+            .ToList();
+        var callResult = this.DefineRegister(node.TypeRequired);
+        var proc = this.TranslateFunctionSymbol(node.Method);
+        if (receiver is null)
         {
-            var args = node.Arguments
-                .Zip(node.Method.Parameters)
-                .Select(pair => this.BoxIfNeeded(pair.Second.Type, this.Compile(pair.First)))
-                .ToList();
-            var callResult = this.DefineRegister(node.TypeRequired);
-            var proc = this.TranslateFunctionSymbol(node.Method);
             this.Write(Call(callResult, proc, args));
-            return isSetter ? args[^1] : callResult;
         }
         else
         {
-            IOperand receiver;
-            // Box receiver, if needed
-            if (node.Method.ContainingSymbol is TypeSymbol methodContainer
-             && NeedsBoxing(methodContainer, node.Receiver.TypeRequired))
-            {
-                receiver = this.Compile(node.Receiver);
-                var newReceiver = this.DefineRegister(methodContainer);
-                this.Write(Box(newReceiver, methodContainer, receiver));
-                receiver = newReceiver;
-            }
-            else
-            {
-                receiver = node.Receiver.TypeRequired.IsValueType
-                    ? this.CompileToAddress(node.Receiver)
-                    : this.Compile(node.Receiver);
-            }
-            var args = node.Arguments
-                .Zip(node.Method.Parameters)
-                .Select(pair => this.BoxIfNeeded(pair.Second.Type, this.Compile(pair.First)))
-                .ToList();
-            var callResult = this.DefineRegister(node.TypeRequired);
-            var proc = this.TranslateFunctionSymbol(node.Method);
             this.Write(MemberCall(callResult, proc, receiver, args));
-            return isSetter ? args[^1] : callResult;
+        }
+        return isSetter ? args[^1] : callResult;
+    }
+
+    private IOperand? CompileReceiver(BoundCallExpression call)
+    {
+        if (call.Receiver is null) return null;
+        // Box receiver, if needed
+        if (call.Method.ContainingSymbol is TypeSymbol methodContainer
+         && NeedsBoxing(methodContainer, call.Receiver.TypeRequired))
+        {
+            var valueReceiver = this.Compile(call.Receiver);
+            var receiver = this.DefineRegister(methodContainer);
+            this.Write(Box(receiver, methodContainer, valueReceiver));
+            return receiver;
+        }
+        else
+        {
+            return call.Receiver.TypeRequired.IsValueType
+                ? this.CompileToAddress(call.Receiver)
+                : this.Compile(call.Receiver);
         }
     }
 

--- a/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
@@ -1,5 +1,6 @@
 using System.Collections.Immutable;
 using System.Linq;
+using System.Runtime.CompilerServices;
 using Draco.Compiler.Internal.BoundTree;
 using Draco.Compiler.Internal.OptimizingIr.Model;
 using Draco.Compiler.Internal.Symbols;
@@ -81,6 +82,31 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
         return proc;
     }
 
+    public IOperand BoxIfNeeded(TypeSymbol targetType, IOperand source)
+    {
+        if (source.Type is null) throw new System.ArgumentException("source must be a typed operand", nameof(source));
+
+        var targetIsValueType = targetType.Substitution.IsValueType;
+        var sourceIsValueType = source.Type.Substitution.IsValueType;
+        var needsBox = !targetIsValueType && sourceIsValueType;
+        var needsUnbox = targetIsValueType && !sourceIsValueType;
+
+        if (needsBox)
+        {
+            var result = this.DefineRegister(targetType.Substitution);
+            this.Write(Box(result, targetType.Substitution, source));
+            return result;
+        }
+
+        if (needsUnbox)
+        {
+            // TODO
+            throw new System.NotImplementedException();
+        }
+
+        return source;
+    }
+
     // Statements //////////////////////////////////////////////////////////////
 
     public override IOperand VisitSequencePointStatement(BoundSequencePointStatement node)
@@ -102,6 +128,7 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
         if (node.Value is null) return default!;
 
         var right = this.Compile(node.Value);
+        right = this.BoxIfNeeded(node.Local.Type, right);
         var left = this.DefineLocal(node.Local);
         this.Write(Store(left, right));
 
@@ -229,9 +256,13 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
     public override IOperand VisitCallExpression(BoundCallExpression node)
     {
         var isSetter = node.Method is IPropertyAccessorSymbol p && p.Property.Setter == node.Method;
+        // TODO: Lots of duplication, we should merge these instructions...
         if (node.Receiver is null)
         {
-            var args = node.Arguments.Select(this.Compile).ToList();
+            var args = node.Arguments
+                .Zip(node.Method.Parameters)
+                .Select(pair => this.BoxIfNeeded(pair.Second.Type, this.Compile(pair.First)))
+                .ToList();
             var callResult = this.DefineRegister(node.TypeRequired);
             var proc = this.TranslateFunctionSymbol(node.Method);
             this.Write(Call(callResult, proc, args));
@@ -242,7 +273,14 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
             var receiver = node.Receiver.TypeRequired.IsValueType
                 ? this.CompileToAddress(node.Receiver)
                 : this.Compile(node.Receiver);
-            var args = node.Arguments.Select(this.Compile).ToList();
+            if (node.Method.ContainingSymbol is TypeSymbol methodContainer)
+            {
+                receiver = this.BoxIfNeeded(methodContainer, receiver);
+            }
+            var args = node.Arguments
+                .Zip(node.Method.Parameters)
+                .Select(pair => this.BoxIfNeeded(pair.Second.Type, this.Compile(pair.First)))
+                .ToList();
             var callResult = this.DefineRegister(node.TypeRequired);
             var proc = this.TranslateFunctionSymbol(node.Method);
             this.Write(MemberCall(callResult, proc, receiver, args));
@@ -339,7 +377,7 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
         }
 
         // Patch
-        PatchStoreSource(leftStore, toStore);
+        this.PatchStoreSource(leftStore, node.Left.Type, toStore);
         this.Write(leftStore);
         return toStore;
     }
@@ -362,8 +400,9 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
         }
     }
 
-    private static void PatchStoreSource(IInstruction storeInstr, IOperand source)
+    private void PatchStoreSource(IInstruction storeInstr, TypeSymbol targetType, IOperand source)
     {
+        source = this.BoxIfNeeded(targetType, source);
         switch (storeInstr)
         {
         case StoreInstruction store:

--- a/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
@@ -1,6 +1,8 @@
 using System.Collections.Immutable;
 using System.Linq;
 using System.Runtime.CompilerServices;
+using Draco.Compiler.Api;
+using Draco.Compiler.Internal.Binding;
 using Draco.Compiler.Internal.BoundTree;
 using Draco.Compiler.Internal.OptimizingIr.Model;
 using Draco.Compiler.Internal.Symbols;
@@ -17,13 +19,16 @@ namespace Draco.Compiler.Internal.OptimizingIr;
 /// </summary>
 internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
 {
+    private readonly Compilation compilation;
     private readonly Procedure procedure;
+
     private BasicBlock currentBasicBlock;
     private bool isDetached;
     private int blockIndex = 0;
 
-    public FunctionBodyCodegen(Procedure procedure)
+    public FunctionBodyCodegen(Compilation compilation, Procedure procedure)
     {
+        this.compilation = compilation;
         this.procedure = procedure;
         // NOTE: Attach block takes care of the null
         this.currentBasicBlock = default!;
@@ -76,7 +81,7 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
         var proc = this.procedure.DeclaringModule.DefineProcedure(func);
         if (!compiledAlready)
         {
-            var codegen = new FunctionBodyCodegen(proc);
+            var codegen = new FunctionBodyCodegen(this.compilation, proc);
             func.Body.Accept(codegen);
         }
         return proc;
@@ -527,6 +532,19 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
 
     public override IOperand VisitFieldExpression(BoundFieldExpression node)
     {
+        // Check, if it's a literal we need to inline
+        var metadataField = ExtractMetadataField(node.Field);
+        if (metadataField is not null && metadataField.IsLiteral)
+        {
+            var defaultValue = metadataField.DefaultValue;
+            if (!BinderFacts.TryGetLiteralType(defaultValue, this.compilation.IntrinsicSymbols, out var literalType))
+            {
+                throw new System.InvalidOperationException();
+            }
+            return new Constant(defaultValue, literalType);
+        }
+
+        // Regular static or nonstatic field
         var receiver = node.Receiver is null ? null : this.Compile(node.Receiver);
         var result = this.DefineRegister(node.TypeRequired);
         this.Write(receiver is null
@@ -534,4 +552,11 @@ internal sealed partial class FunctionBodyCodegen : BoundTreeVisitor<IOperand>
             : LoadField(result, receiver, node.Field));
         return result;
     }
+
+    private static MetadataFieldSymbol? ExtractMetadataField(FieldSymbol field) => field switch
+    {
+        MetadataFieldSymbol m => m,
+        FieldInstanceSymbol i => ExtractMetadataField(i.GenericDefinition),
+        _ => null,
+    };
 }

--- a/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/FunctionBodyCodegen.cs
@@ -1,6 +1,5 @@
 using System.Collections.Immutable;
 using System.Linq;
-using System.Runtime.CompilerServices;
 using Draco.Compiler.Api;
 using Draco.Compiler.Internal.Binding;
 using Draco.Compiler.Internal.BoundTree;

--- a/src/Draco.Compiler/Internal/OptimizingIr/InstructionFactory.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/InstructionFactory.cs
@@ -11,6 +11,8 @@ namespace Draco.Compiler.Internal.OptimizingIr;
 internal static class InstructionFactory
 {
     public static NopInstruction Nop() => new();
+    public static BoxInstruction Box(Register target, TypeSymbol boxedType, IOperand value) =>
+        new(target, boxedType, value);
     public static StoreInstruction Store(IOperand target, IOperand source) => new(target, source);
     public static StoreElementInstruction StoreElement(IOperand array, IEnumerable<IOperand> indices, IOperand source) =>
         new(array, indices, source);

--- a/src/Draco.Compiler/Internal/OptimizingIr/Model/BoxInstruction.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/Model/BoxInstruction.cs
@@ -1,0 +1,41 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Draco.Compiler.Internal.Symbols;
+
+namespace Draco.Compiler.Internal.OptimizingIr.Model;
+
+/// <summary>
+/// Valuetype element boxing.
+/// </summary>
+internal sealed class BoxInstruction : InstructionBase
+{
+    /// <summary>
+    /// The register to write the boxed value to.
+    /// </summary>
+    public Register Target { get; set; }
+
+    /// <summary>
+    /// The boxed type.
+    /// </summary>
+    public TypeSymbol BoxedType { get; set; }
+
+    /// <summary>
+    /// The boxed value.
+    /// </summary>
+    public IOperand Value { get; }
+
+    public BoxInstruction(Register target, TypeSymbol boxedType, IOperand value)
+    {
+        this.Target = target;
+        this.BoxedType = boxedType;
+        this.Value = value;
+    }
+
+    public override string ToString() =>
+        $"{this.Target.ToOperandString()} := box {this.Value.ToOperandString()} as {this.BoxedType}";
+
+    public override BoxInstruction Clone() => new(this.Target, this.BoxedType, this.Value);
+}

--- a/src/Draco.Compiler/Internal/OptimizingIr/Model/BoxInstruction.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/Model/BoxInstruction.cs
@@ -1,8 +1,3 @@
-using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
 using Draco.Compiler.Internal.Symbols;
 
 namespace Draco.Compiler.Internal.OptimizingIr.Model;

--- a/src/Draco.Compiler/Internal/OptimizingIr/ModuleCodegen.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/ModuleCodegen.cs
@@ -23,7 +23,7 @@ internal sealed class ModuleCodegen : SymbolVisitor
     {
         this.compilation = compilation;
         this.module = module;
-        this.globalInitializer = new(module.GlobalInitializer);
+        this.globalInitializer = new(this.compilation, module.GlobalInitializer);
         this.emitSequencePoints = emitSequencePoints;
     }
 
@@ -70,7 +70,7 @@ internal sealed class ModuleCodegen : SymbolVisitor
         // Yank out potential local functions and closures
         var (bodyWithoutLocalFunctions, localFunctions) = ClosureRewriter.Rewrite(body);
         // Compile it
-        var bodyCodegen = new FunctionBodyCodegen(procedure);
+        var bodyCodegen = new FunctionBodyCodegen(this.compilation, procedure);
         bodyWithoutLocalFunctions.Accept(bodyCodegen);
 
         // Compile the local functions

--- a/src/Draco.Compiler/Internal/OptimizingIr/ModuleCodegen.cs
+++ b/src/Draco.Compiler/Internal/OptimizingIr/ModuleCodegen.cs
@@ -49,6 +49,7 @@ internal sealed class ModuleCodegen : SymbolVisitor
             // Compile it
             var value = bodyWithoutLocalFunctions.Accept(this.globalInitializer);
             // Store it
+            value = this.globalInitializer.BoxIfNeeded(global.Type, value);
             this.globalInitializer.Write(Store(global, value));
 
             // Compile the local functions

--- a/src/Draco.Compiler/Internal/Solver/ConstraintSolver.cs
+++ b/src/Draco.Compiler/Internal/Solver/ConstraintSolver.cs
@@ -70,7 +70,7 @@ internal sealed partial class ConstraintSolver
             var unwrappedLocalType = localType.Substitution;
             if (unwrappedLocalType is TypeVariable typeVar)
             {
-                this.Unify(typeVar, IntrinsicSymbols.UninferredType);
+                this.UnifyAsserted(typeVar, IntrinsicSymbols.UninferredType);
                 diagnostics.Add(Diagnostic.Create(
                     template: TypeCheckingErrors.CouldNotInferType,
                     location: local.DeclaringSyntax.Location,
@@ -98,7 +98,7 @@ internal sealed partial class ConstraintSolver
         foreach (var typeVar in this.typeVariables)
         {
             var unwrapped = typeVar.Substitution;
-            if (unwrapped is TypeVariable unwrappedTv) this.Unify(unwrappedTv, IntrinsicSymbols.UninferredType);
+            if (unwrapped is TypeVariable unwrappedTv) this.UnifyAsserted(unwrappedTv, IntrinsicSymbols.UninferredType);
         }
     }
 
@@ -184,12 +184,23 @@ internal sealed partial class ConstraintSolver
         this.substitutions.Add(var, type);
 
     /// <summary>
+    /// Unifies two types, asserting their success.
+    /// </summary>
+    /// <param name="first">The first type to unify.</param>
+    /// <param name="second">The second type to unify.</param>
+    public void UnifyAsserted(TypeSymbol first, TypeSymbol second)
+    {
+        if (this.Unify(first, second)) return;
+        throw new System.InvalidOperationException($"could not unify {first} and {second}");
+    }
+
+    /// <summary>
     /// Attempts to unify two types.
     /// </summary>
     /// <param name="first">The first type to unify.</param>
     /// <param name="second">The second type to unify.</param>
     /// <returns>True, if unification was successful, false otherwise.</returns>
-    public bool Unify(TypeSymbol first, TypeSymbol second)
+    private bool Unify(TypeSymbol first, TypeSymbol second)
     {
         first = first.Substitution;
         second = second.Substitution;

--- a/src/Draco.Compiler/Internal/Solver/ConstraintSolver_Rules.cs
+++ b/src/Draco.Compiler/Internal/Solver/ConstraintSolver_Rules.cs
@@ -335,9 +335,14 @@ internal sealed partial class ConstraintSolver
                     this.UnifyParameterWithArgument(param.Type, arg);
                 }
             }
-            // NOTE: Unification won't always be correct, especially not when subtyping arises
-            // In all cases, return type is simple
-            this.UnifyAsserted(constraint.ReturnType, chosen.ReturnType);
+            // In all cases, return type is simple, it's an assignment
+            var returnTypePromise = this.Assignable(constraint.ReturnType, chosen.ReturnType);
+            // TODO: This location config is horrible, we need that refactor SOON
+            if (constraint.Diagnostic.Location is not null)
+            {
+                returnTypePromise.ConfigureDiagnostic(diag => diag
+                    .WithLocation(constraint.Diagnostic.Location));
+            }
             // Resolve promise
             constraint.Promise.Resolve(chosen);
         }

--- a/src/Draco.Compiler/Internal/Solver/ConstraintSolver_Rules.cs
+++ b/src/Draco.Compiler/Internal/Solver/ConstraintSolver_Rules.cs
@@ -89,9 +89,8 @@ internal sealed partial class ConstraintSolver
                 .ToList();
             if (assignmentsWithSameTarget.Count == 0)
             {
-                // TODO: Unconditional unification...
                 // No, assume same type
-                this.Unify(assignable.TargetType, assignable.AssignedType);
+                this.UnifyAsserted(assignable.TargetType, assignable.AssignedType);
                 return true;
             }
 
@@ -126,7 +125,7 @@ internal sealed partial class ConstraintSolver
             // NOTE: We do NOT remove the constraint, will be resolved in a future iteration
             // Only one non-type-var, the rest are type variables
             var nonTypeVar = commonNonTypeVars[0];
-            foreach (var tv in commonTypeVars) this.Unify(tv, nonTypeVar);
+            foreach (var tv in commonTypeVars) this.UnifyAsserted(tv, nonTypeVar);
             return true;
         }
 
@@ -193,7 +192,7 @@ internal sealed partial class ConstraintSolver
             if (constraint.AlternativeTypes.All(t => SymbolEqualityComparer.Default.IsBaseOf(type, t)))
             {
                 // Found a good common type
-                this.Unify(constraint.CommonType, type);
+                this.UnifyAsserted(constraint.CommonType, type);
                 constraint.Promise.Resolve(default);
                 return;
             }
@@ -203,7 +202,7 @@ internal sealed partial class ConstraintSolver
         constraint.Diagnostic
             .WithTemplate(TypeCheckingErrors.NoCommonType)
             .WithFormatArgs(string.Join(", ", constraint.AlternativeTypes));
-        this.Unify(constraint.CommonType, IntrinsicSymbols.ErrorType);
+        this.UnifyAsserted(constraint.CommonType, IntrinsicSymbols.ErrorType);
         constraint.Promise.Fail(default, diagnostics);
     }
 
@@ -229,7 +228,7 @@ internal sealed partial class ConstraintSolver
                 .WithFormatArgs(constraint.MemberName, accessed);
             // We still provide a single error symbol
             var errorSymbol = new UndefinedMemberSymbol();
-            this.Unify(constraint.MemberType, IntrinsicSymbols.ErrorType);
+            this.UnifyAsserted(constraint.MemberType, IntrinsicSymbols.ErrorType);
             constraint.Promise.Fail(errorSymbol, diagnostics);
             return;
         }
@@ -237,7 +236,7 @@ internal sealed partial class ConstraintSolver
         if (membersWithName.Length == 1)
         {
             // One member, we know what type the member type is
-            this.Unify(((ITypedSymbol)membersWithName[0]).Type, constraint.MemberType);
+            this.UnifyAsserted(((ITypedSymbol)membersWithName[0]).Type, constraint.MemberType);
             constraint.Promise.Resolve(membersWithName[0]);
             return;
         }
@@ -247,7 +246,7 @@ internal sealed partial class ConstraintSolver
             // All must be functions, otherwise we have bigger problems
             // TODO: Can this assertion fail? Like in a faulty module decl?
             Debug.Assert(membersWithName.All(m => m is FunctionSymbol));
-            this.Unify(constraint.MemberType, IntrinsicSymbols.ErrorType);
+            this.UnifyAsserted(constraint.MemberType, IntrinsicSymbols.ErrorType);
             var overload = new OverloadSymbol(membersWithName.Cast<FunctionSymbol>().ToImmutableArray());
             constraint.Promise.Resolve(overload);
         }
@@ -293,7 +292,7 @@ internal sealed partial class ConstraintSolver
         // We have all candidates well-defined, find the absolute dominator
         if (candidates.Count == 0)
         {
-            this.Unify(constraint.ReturnType, IntrinsicSymbols.ErrorType);
+            this.UnifyAsserted(constraint.ReturnType, IntrinsicSymbols.ErrorType);
             // Best-effort shape approximation
             var errorSymbol = new NoOverloadFunctionSymbol(constraint.Arguments.Length);
             constraint.Diagnostic
@@ -338,14 +337,14 @@ internal sealed partial class ConstraintSolver
             }
             // NOTE: Unification won't always be correct, especially not when subtyping arises
             // In all cases, return type is simple
-            this.Unify(constraint.ReturnType, chosen.ReturnType);
+            this.UnifyAsserted(constraint.ReturnType, chosen.ReturnType);
             // Resolve promise
             constraint.Promise.Resolve(chosen);
         }
         else
         {
             // Best-effort shape approximation
-            this.Unify(constraint.ReturnType, IntrinsicSymbols.ErrorType);
+            this.UnifyAsserted(constraint.ReturnType, IntrinsicSymbols.ErrorType);
             var errorSymbol = new NoOverloadFunctionSymbol(constraint.Arguments.Length);
             constraint.Diagnostic
                 .WithTemplate(TypeCheckingErrors.AmbiguousOverloadedCall)
@@ -374,7 +373,7 @@ internal sealed partial class ConstraintSolver
         if (called is not FunctionTypeSymbol functionType)
         {
             // Error
-            this.Unify(constraint.ReturnType, IntrinsicSymbols.ErrorType);
+            this.UnifyAsserted(constraint.ReturnType, IntrinsicSymbols.ErrorType);
             constraint.Diagnostic
                 .WithTemplate(TypeCheckingErrors.CallNonFunction)
                 .WithFormatArgs(called);
@@ -384,13 +383,13 @@ internal sealed partial class ConstraintSolver
 
         // It's a function
         // We can merge the return type
-        this.Unify(constraint.ReturnType, functionType.ReturnType);
+        this.UnifyAsserted(constraint.ReturnType, functionType.ReturnType);
 
         // Check if it has the same number of args
         if (functionType.Parameters.Length != constraint.Arguments.Length)
         {
             // Error
-            this.Unify(constraint.ReturnType, IntrinsicSymbols.ErrorType);
+            this.UnifyAsserted(constraint.ReturnType, IntrinsicSymbols.ErrorType);
             constraint.Diagnostic
                 .WithTemplate(TypeCheckingErrors.TypeMismatch)
                 .WithFormatArgs(
@@ -408,7 +407,7 @@ internal sealed partial class ConstraintSolver
             if (score.HasZero)
             {
                 // Error
-                this.Unify(constraint.ReturnType, IntrinsicSymbols.ErrorType);
+                this.UnifyAsserted(constraint.ReturnType, IntrinsicSymbols.ErrorType);
                 constraint.Diagnostic
                     .WithTemplate(TypeCheckingErrors.TypeMismatch)
                     .WithFormatArgs(
@@ -430,14 +429,14 @@ internal sealed partial class ConstraintSolver
 
     private void FailRule(OverloadConstraint constraint)
     {
-        this.Unify(constraint.ReturnType, IntrinsicSymbols.ErrorType);
+        this.UnifyAsserted(constraint.ReturnType, IntrinsicSymbols.ErrorType);
         var errorSymbol = new NoOverloadFunctionSymbol(constraint.Arguments.Length);
         constraint.Promise.Fail(errorSymbol, null);
     }
 
     private void FailRule(CallConstraint constraint)
     {
-        this.Unify(constraint.ReturnType, IntrinsicSymbols.ErrorType);
+        this.UnifyAsserted(constraint.ReturnType, IntrinsicSymbols.ErrorType);
         constraint.Promise.Fail(default, null);
     }
 }

--- a/src/Draco.Compiler/Internal/Symbols/Metadata/MetadataFieldSymbol.cs
+++ b/src/Draco.Compiler/Internal/Symbols/Metadata/MetadataFieldSymbol.cs
@@ -49,6 +49,18 @@ internal sealed class MetadataFieldSymbol : FieldSymbol, IMetadataSymbol
     /// </summary>
     public MetadataReader MetadataReader => this.Assembly.MetadataReader;
 
+    /// <summary>
+    /// True, if this is a literal that cannot be referenced as a field, but needs to be inlined as a value.
+    /// This is the case for enum members.
+    /// </summary>
+    public bool IsLiteral => this.fieldDefinition.Attributes.HasFlag(FieldAttributes.Literal);
+
+    /// <summary>
+    /// The default value of this field.
+    /// </summary>
+    public object? DefaultValue => InterlockedUtils.InitializeMaybeNull(ref this.defaultValue, this.BuildDefaultValue);
+    private object? defaultValue;
+
     private readonly FieldDefinition fieldDefinition;
 
     public MetadataFieldSymbol(Symbol containingSymbol, FieldDefinition fieldDefinition)
@@ -62,5 +74,14 @@ internal sealed class MetadataFieldSymbol : FieldSymbol, IMetadataSymbol
         // Decode signature
         var decoder = new TypeProvider(this.Assembly.Compilation);
         return this.fieldDefinition.DecodeSignature(decoder, this);
+    }
+
+    private object? BuildDefaultValue()
+    {
+        var constantHandle = this.fieldDefinition.GetDefaultValue();
+        if (constantHandle.IsNil) return null;
+
+        var constant = this.MetadataReader.GetConstant(constantHandle);
+        return MetadataSymbol.DecodeConstant(constant, this.MetadataReader);
     }
 }

--- a/src/Draco.Compiler/Internal/Symbols/Metadata/MetadataSymbol.cs
+++ b/src/Draco.Compiler/Internal/Symbols/Metadata/MetadataSymbol.cs
@@ -1,3 +1,4 @@
+using System;
 using System.Collections.Generic;
 using System.Reflection;
 using System.Reflection.Metadata;
@@ -73,6 +74,35 @@ internal static class MetadataSymbol
             };
         }
         return null;
+    }
+
+    public static object? DecodeConstant(Constant constant, MetadataReader metadataReader)
+    {
+        var blob = metadataReader.GetBlobBytes(constant.Value);
+        return constant.TypeCode switch
+        {
+            ConstantTypeCode.NullReference => null,
+
+            ConstantTypeCode.Boolean => BitConverter.ToBoolean(blob),
+            ConstantTypeCode.Char => BitConverter.ToChar(blob),
+
+            ConstantTypeCode.String => BitConverter.ToString(blob),
+
+            ConstantTypeCode.Byte => blob[0],
+            ConstantTypeCode.UInt16 => BitConverter.ToUInt16(blob),
+            ConstantTypeCode.UInt32 => BitConverter.ToUInt32(blob),
+            ConstantTypeCode.UInt64 => BitConverter.ToUInt64(blob),
+
+            ConstantTypeCode.SByte => (sbyte)blob[0],
+            ConstantTypeCode.Int16 => BitConverter.ToInt16(blob),
+            ConstantTypeCode.Int32 => BitConverter.ToInt32(blob),
+            ConstantTypeCode.Int64 => BitConverter.ToInt64(blob),
+
+            ConstantTypeCode.Single => BitConverter.ToSingle(blob),
+            ConstantTypeCode.Double => BitConverter.ToDouble(blob),
+
+            _ => throw new ArgumentOutOfRangeException(nameof(constant)),
+        };
     }
 
     private static FunctionSymbol SynthetizeConstructor(

